Index: commands.pl
===================================================================
--- commands.pl	(révision 430)
+++ commands.pl	(copie de travail)
@@ -828,6 +828,12 @@
 	    return 1;
     }
 
+    # Quick and dirty hack: forbid self-votes in day phase
+    if($phs{'current'} eq 'day' && $ni eq $voted_for_ni) {
+       say(P_GAME, 'error', $ni, "Vous ne pouvez pas voter contre vous-même.");
+       return 1;
+    }
+
     # Check if the vote is the same as last one
     if(defined($old_vote) && $old_vote eq $voted_for_ni) {
 	&$reply(merge_votemsg($messages{'votes'}{'vote'}),
